networkx/drawing/nx_pylab.py

```diff
@@ -51,6 +53,1091 @@
+            )
+
+    if canvas is None:
+        canvas = plt.gca()
+
+    if kwargs.get("hide_ticks", defaults["hide_ticks"]):
+        canvas.tick_params(
+            axis="both",
+            which="both",
+            bottom=False,
+            left=False,
+            labelbottom=False,
+            labelleft=False,
+        )
+
+    ### Helper methods and classes
+
+    # I hope that the order-perserving nature of dicts in python 3.7+
+    # means that G.nodes() will always return the same order.
+    def node_property_sequence(seq, attr, default=None):
+        """Return a sequence of attribute values for the given sequence, using default if not None"""
+
+        # If the attr isn't actually a graph attr, but was explicitly passed as an argument it must be a
+        # user-default value. Allow None to be used to tell draw to skip attributes which are on the graph
+        if (
+            attr is not None
+            and nx.get_node_attributes(node_subgraph, attr) == {}
+            and any(attr == v for k, v in kwargs.items() if "node" in k)
+        ):
+            return [attr for _ in seq]
+
+        # The inner function isn't my first choice here, but if I use a regular try-except
+        # around the asarray call I can't seem to isolate which node is problematic for
+        # the error message.
+        def error(n, attr, default):
+            if default is None:
+                raise nx.NetworkXError(
+                    f"Node {n} is missing required attribute: {attr}"
+                )
+            else:
+                return False
+
+        return [
+            (
+                node_subgraph.nodes[n][attr]
+                if attr in node_subgraph.nodes[n] or error(n, attr, default)
+                else default
+            )
+            for n in seq
+        ]
+
+    def compute_colors(color, alpha):
+        if isinstance(color, str):
+            rgba = mpl.colors.colorConverter.to_rgba(color)
+            if (
+                alpha != defaults["node_alpha"]
+            ):  # Using a non-default alpha value overrides any alpha value in the color
+                return (rgba[0], rgba[1], rgba[2], alpha)
+            return rgba
+
+        if isinstance(color, tuple) and len(color) == 3:
+            return (color[0], color[1], color[2], alpha)
+
+        if isinstance(color, tuple) and len(color) == 4:
+            return color
+
+        raise ValueError(f"Invalid format for color: {color}")
+
+    # Find which edges can be plotted as a line collection
+    #
+    # Non-default values for these attributes require fancy arrow patches:
+    # - any arrow style (including the default -|> for directed graphs)
+    # - arrow size (by extension of style)
+    # - connection style
+    # - min_source_margin
+    # - min_target_margin
+
+    def collection_compatible(e):
+        return (
+            get_edge_attr(
+                e, kwargs.get("arrowstyle", "arrowstyle"), defaults["arrowstyle"]
+            )
```

Comment:
This seems like a lot of boilerplate that gets in the way of clarity.  Can we make the `get_edge_attr` function handle the defaults and kwargs get cruft?  It should have access to the kwargs and defaults variables from the function namespace, right?

Reply:
I agree. I did have to make some changes to the function parameter names to get this working. Basically, the two arguments to `get` aren't always the same (since I typically strip off the leading `node_` or `edge_` so that `node_color` becomes a node attribute of `color`). To keep everything consistent, I added the below prefixes:
- `pos` to `node_pos`
- `arrowstyle` to `edge_arrowstyle`
- `arrowsize` to `edge_arrowsize`